### DISCUSSION TITLE
`mutate` callbacks can fire after unmount

### DISCUSSION THREAD

**Discussion by janlat** [TanStack/query#4804] [Q&A] [answered]
According to [the docs](https://react-query-v3.tanstack.com/guides/mutations#mutation-side-effects)
>additional callbacks won't run if your component unmounts before the mutation finishes

however this is only true if the mutation was already in progress during the unmount. If you start the mutation after the unmount, they will run. Is this intentional?

**Comment by TkDodo** [comment]
can you show a codesandbox example of what you mean?

**Reply by janlat** [reply]
https://codesandbox.io/s/keen-voice-n60pe3

**Reply by TkDodo** [reply]
ah yes, that's a good point, I can explain why this is happening: With `useMutation`, an `observer` is only added when `mutate` is invoked. `observers` are removed when the component unmounts to avoid the callbacks from running.

Now if you add the observer after the component has unmounted (which you do by calling mutate), there is no further even that stops the callbacks from running (The query logic lives outside of react). I think I've never invoked a function from a react component after the component has unmounted.

**Reply by janlat** [reply]
Yeah, it's not common, but possible nevertheless :stuck_out_tongue:

I guess it would be quite easy to fix [in v3](https://github.com/TanStack/query/blob/5e37b145a8836b00dd20860e8eb1f94c608fec81/src/react/useMutation.ts#L114) by ignoring passed callbacks if `mountedRef.current` is `false`. I'm not sure about [v4](https://github.com/TanStack/query/blob/e33bda3131442f34577f13c73f3c6e890b504a62/packages/react-query/src/useMutation.ts) though, is it even handled there? :thinking:

**Reply by TkDodo** [reply]
we don't have the "mountedRef" tracking anymore in v4, no.

**Reply by janlat** [reply]
Actually, it probably would be better to change
https://github.com/TanStack/query/blob/e33bda3131442f34577f13c73f3c6e890b504a62/packages/query-core/src/mutationObserver.ts#L133
into
```js
if (this.listeners.length) {
  this.currentMutation.addObserver(this)
}
```

**Reply by TkDodo** [reply]
Hmm, could you maybe create a PR that has a failing test case with the current implementation, so that we can see if this would fix the issue without introducing other regressions?

**Reply by janlat** [reply]
https://github.com/TanStack/query/pull/4848

**Reply by louis-young** [reply]
Hi 👋 

Thanks for writing the test. I'm not a maintainer but I've verified the failing test case and tested the proposed fix. It seems to fix the issue without introducing any regressions. The change seems sensible: we won't add an observer for the mutation if there are no listeners (i.e we won't observe `mutate` callbacks if the component has been unmounted before mutating).

I can see there's a related test case here, although it unmounts the component _after_ the mutation has started, not before:

https://github.com/TanStack/query/blob/431fc09eccc524507a3e39ffcd35b78aeeb4db5e/packages/react-query/src/__tests__/useMutation.test.tsx#L848)

Perhaps it'd be worth updating this test case to be more explicit at the same time, something like this:

```diff
- 'should call cache callbacks when unmounted'
+ 'should only call cache callbacks when unmounted after mutating'
```

It also might be worth updating the new test case to assert that the cache callbacks are still called?

I'll leave this to Dominik but it looks good to me! Thanks again 🚀

**Comment by louis-young** [comment, marked as answer]
This is/will be resolved by https://github.com/TanStack/query/pull/4848. Thanks everyone 🚀
